[Swiftify] Always remove count parameters when possible

hnrklssn opened the pull request:
Previously we did not remove count parameters if any count parameters were shared between count expressions, or if any count expression contained operations. Buffer sizes were also just checked to be larger than or equal than the given count.

We now extract the count from Spans/BufferPointers whenever possible, and store that value in a variable at the start of the function. If multiple parameters share the same count, a bounds check is emitted to make sure that they have the same size. Subspans can be used if one span is larger than necessary.

The message in the bounds check is changed so that it includes the expected and actual value, to aid in debugging.

This patch also fixes some incorrect indentation, and adds the Whitespace.swift test case to act as a regression test in case the indentation changes, since the other test cases don't use significant whitespace.

rdar://151488820
rdar://151511090
rdar://146333006
rdar://147715799

hnrklssn committed:
[Swiftify] Always remove count parameters when possible

Previously we did not remove count parameters if any count parameters
were shared between count expressions, or if any count expression
contained operations. Buffer sizes were also just checked to be larger
than or equal than the given count.

We now extract the count from Spans/BufferPointers whenever possible,
and store that value in a variable at the start of the function. If
multiple parameters share the same count, a bounds check is emitted to
make sure that they have the same size. Subspans can be used if one span
is larger than necessary.

The message in the bounds check is changed so that it includes the
expected and actual value, to aid in debugging.

This patch also fixes some incorrect indentation, and adds the
Whitespace.swift test case to act as a regression test in case the
indentation changes, since the other test cases don't use significant
whitespace.

rdar://151488820
rdar://151511090
rdar://146333006
rdar://147715799

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

DougGregor reviewed (approved):
This is a fantastic clean-up, thank you!

DougGregor commented:
@swift-ci please smoke test Windows

hnrklssn committed:
[Swiftify] Add tests that verify __counted_by with constant count (NFC)

We used to not emit bounds checks for things like `__counted_by(4)`, but
we do now - likely fixed by
https://github.com/swiftlang/swift/pull/81585
These regression tests verify that behaviour.

rdar://151038254
